fendermint/vm/interpreter/src/chain.rs

```diff
@@ -221,32 +238,51 @@ where
                     }
 
                     // We can use the same result type for now, it's isomorphic.
-                    Ok(((pool, state), ChainMessageApplyRet::Signed(ret)))
+                    Ok(((pool, provider, state), ChainMessageApplyRet::Signed(ret)))
                 }
                 IpcMessage::BottomUpExec(_) => {
                     todo!("#197: implement BottomUp checkpoint execution")
                 }
-                IpcMessage::TopDownProposal(_) => {
-                    todo!("implement TopDown handling; this is just a placeholder")
+                IpcMessage::TopDownProposal(p) => {
+                    let validator_set = downcast_err!(
+                        atomically_or_err(|| provider.validator_set(p.height as u64)).await
+                    )?
+                    .ok_or_else(|| anyhow!("cannot find validator set for block: {}", p.height))?;
```

Comment:
I think this one can be tricky as well. `ipc-agent` is an external service, even when we query it, it could be down or its data is somehow lagging. Then we still facing the same problem now. Do we continue execution or we stop the node to check what happened? I assumed (which I could be wrong) it's part of the service level agreement that parent finality provider should contain the data for execution, if it does not, something is worth checking.